velox/expression/EvalCtx.cpp

```diff
@@ -273,6 +273,49 @@ VectorPtr EvalCtx::ensureFieldLoaded(
   return field;
 }
 
+namespace {
+// Helper function for addNulls that ensures result is unique and have writable
+// nulls for size.
+// The method should also guarantee that result is not constant since addNulls
+// is not defined for constant vectors.
+void prepareForNullWriting(
```

Comment:
I feel like the logic is very fragmented between this and addNulls() which makes it difficult to read through the full flow of addNulls and be confident that we covered everything. Can we structure the original addNulls as following:

1. result is NULL
2. Constant encoding:
  - Null Const: resize if required
  - Non-Null : wrap in dict layer (possible optimizations include setting dict idx = const idx)
3. Dict
  - unique check then modify directly ELSE create a new dict layer (no need to wrap on top just replace the existing layer)
4. FlatVector :
  - unique check + ensureWritable + add Nulls
5. Complex:
  - Wrap in dict layer